README.md
# IPK Projekt 1: Protokol IPK Kalkulačky

Mé řešení prvního projektu je psaný v jazyce C++ a sestává ze 3 .cpp souborů, tcp.cpp, udp.cpp a ipkcpc.cpp a 2 hlavičkových souborů tcp.h a udp.h.
# TCP 
V souboru tcp.cpp je vypracováno řešení tcp komunikace se serverem. Primární inspirace vzešla ze studijních materiálů, ale nevázala se k nim žádná licence. Správný postup by byl kontaktovat autora a domluvit se sním o využití jeho díla, ale tato akce mi přišla nevhodná vzhledem k počtu studentů, kteří by mohli mít stejný nápad, a pak by pan docent Ryšavý obdržel nespočet mailů ohledně jednoho kódu. <br>
Program funguje na principu navázání spojení a neustále pravdivého while cyklu, který se přeruší po vypsání BYE anebo chybě. Klient poté komunikaci se serverem ukončí. Pokud dojde k chybě nebo neočekávanému ukončení spojení, klient sám pošle BYE serveru a čeká na obdržení téže zprávy zpět, poté bezpečně spojení ukončí. <br>
# UDP
Soubor udp.cpp je principielně stejný jako tcp.cpp s tím rozdílem, že se data při posílání posouvají o 2 byty, které se také samozřejmě zpracují a při příjímání nastává posuv o 3 byty. Ukončení spojení se vykoná koncem soubodu, neboli EOF (popř. vyvolanou chybou).<br>
# Hlavičkové soubory
Hlavičkové soubory tcp.h a udp.h slouží pro lepší přehlednost a ačkoliv jsou v nich deklarovány pouze 2 samostatné funkce, myslím si, že je vhodné se s nimi naučit pracovat pro složitější programy. Díky těmto souborům je tak mnohem jednodušší jejich použití v hlavním ipkcpc.cpp programu.
# IPKCPC.CPP
Tento soubor slouží pro načtení a kontrolu argumentů a poté jejich spuštění na základě jejich parametrů. Pomocí parametru --help vypíše popis povinných argumentů pro správné spuštění programu. Argumenty jsou podle specifikace zkontrolovány a podle zvoleného módu je spuštěno spojení se serverem. <br>
## Makefile
Program je přeložitelný pomocí příkazu *make* a spustitelný klasickým způsobem s potřebnými argumenty.
## Licence
Program je dostupný pod licencí MIT, která je popsána v souboru *LICENSE*.
## Testování
Program prošel ručním testování na dvou Unixových zařízení, WSL Ubuntu a Fedora. Oba tyto systémy jsou používány jako virtuální stroje. <br>
**Specifikace stroje:**
![Specifikace zařízení - Fedora](/ss/Fedora_screen.png "Fedora") <br>

Nejprve byl testován klasický překlad a spuštění, tj. připojení na server a následné odpojení. Kontrola správnosti probíhala pomoci aplikace Wireshark následovně: <br>
![Ukázka použití aplikace Wireshark](/ss/Wireshark.png "Wireshark") <br>
Testování proběhlo na obou strojích obdobně pro tcp i udp a výsledky byly totožné.
Otestovány byly i chybné vstupy a nečekané ukončení spojení a proces se vždy dokázal v pořádku dokončit nebo v případně špatných parametrů vůbec nespustit.<br>
V neposlední řadě byl program otestován pomocí aplikace valgrind, díky čemuž byly odhaleny paměťové úniky a mohly být následně opraveny.<br>
Valgrind výsledky:
![Ukázka testování pomocí Valgrind](/ss/Valgrind_UDP.png "Valgrind UDP") <br>

## Zdroje
Jak je zmíněno výše, primární zdoj byly studijní materiály a přednášky. Pro pomoc s jazykem C++ jsem využila stránku https://www.tutorialspoint.com/index.htm, kde se nachází velké množství rad a dokumentace k funkcím jazyka.<br>

## Dokumentace
Pro popis programu byla využita služba Doxygen, pomocí které jsou popsány veškeré části všech souborů.


CHANGELOG.md
# Limity a problémy programu
Program nebyl testován na Windowsovém zařízení, není tudíž 100% jistá jeho multiplatformní kompatibilita. Implementace projektu byla směřována na Unixová zařízení. Tato situace se dá vyřešit například při překladu, kde by Makefile zjistil, na kterém zařízení se překlad vyvolal a na základě toho spouštěl odlišné části programu. V mém řešení se nachází funkce bzero(), která není kompatibilní s Windows zařízeními a tudíž by se na těchto místech vyskytl problém. Řešením by mohlo být například využití fuknce ZeroMemory(), která je vhodnější pro Windows. Dále by se musely využít jiné knihovny a funkce.<br> <br>
Následovně se v programu nachází omezení velikosti bufferu na 1024 znaků, tudíž by při přesáhnutí této hodnoty vznikl problém. Implementace většího bufferu mi nepřišla potřebná, ale tuto limitaci chci zaznamenat.<br>
Testování probíhalo ručně a pouze několika pokusy, není tedy dostatečné a všechny nedostatky tudíž nejsou známy. 

tcp.cpp
/**
 * @file        tcp.cpp
 * @author      Veronika Jirmusova, xjirmu00
 * @brief       TCP communication with server
 * @version     0.1
 * @date        2023-03-15
 * 
 * @copyright Copyright (c) 2023
 * 
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <unistd.h>
#include <iostream>
using namespace std;
#define BUFSIZE 1024

/**
 * @brief   Main TCP function   
 * 
 * @param   port_number         wanted port number
 * @param   server_hostname     name of server
 * @return int 
 */
int tcp(int port_number, char *server_hostname){
    
    int client_socket, bytestx, bytesrx;
    socklen_t serverlen;
    struct hostent *server;
    struct sockaddr_in server_address;
    char buf[BUFSIZE];

    //Check if the name of wanted server is correct
    if ((server = gethostbyname(server_hostname)) == NULL){
            fprintf(stderr,"ERROR: no such host as %s\n", server_hostname);
            exit(EXIT_FAILURE);
        }
        
        //Find an IP adress of wanted server
        bzero((char *) &server_address, sizeof(server_address));
        server_address.sin_family = AF_INET;
        bcopy((char *)server->h_addr, (char *)&server_address.sin_addr.s_addr, server->h_length);
        server_address.sin_port = htons(port_number);
        
        //Print info about the server socket
        //printf("INFO: Server socket: %s : %d \n", inet_ntoa(server_address.sin_addr), ntohs(server_address.sin_port));
        
        //Declaration of socket
        if ((client_socket = socket(AF_INET, SOCK_STREAM, 0)) <= 0)
        {
            perror("ERROR: socket");
            exit(EXIT_FAILURE);
        }
        
        ///Get the message from user
        if (connect(client_socket, (const struct sockaddr *) &server_address, sizeof(server_address)) != 0)
        {
            perror("ERROR: connect");
            exit(EXIT_FAILURE);        
        }

    bool bye = false;
    //Established connection
    while(true){
        bzero(buf, BUFSIZE);
        //printf("Please enter msg: ");
        //Send BYE to server if an error is caught
        if(fgets(buf, BUFSIZE, stdin) == NULL){
            strcpy(buf, "BYE\n");
            send(client_socket, buf, strlen(buf), 0);
            recv(client_socket, buf, BUFSIZE, 0);
            break;
        }
            
        //Send a message to server
        bytestx = send(client_socket, buf, strlen(buf), 0);
        if (bytestx < 0) 
            perror("ERROR in sendto");
        
        bzero(buf, BUFSIZE);

        //Receive an answer
        bytesrx = recv(client_socket, buf, BUFSIZE, 0);
        if (bytesrx < 0) 
            perror("ERROR in recvfrom");
        
        printf("%s", buf);

        if(!strcmp(buf, "BYE\n")){
            break;
        }
    }
    close(client_socket);
    return 0;
}
//End Of File tcp.cpp

udp.cpp
/**
 * @file        udp.cpp
 * @author      Veronika Jirmusova, xjirmu00
 * @brief       UCP connection
 * @version     0.1
 * @date        2023-03-18
 * 
 * @copyright   Copyright (c) 2023
 * 
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <iostream>
#include <unistd.h>
using namespace std;
#define BUFSIZE 1024

/**
 * @brief   Move bytes in buffer when sending message
 * 
 * @param   buffer      buffer with data
 */
void iLikeToMoveItMoveIt(char *buffer){
    size_t len = strlen(buffer);
    for (int i = BUFSIZE - 1; i > 1; i--) {
        buffer[i] = buffer[i - 2];
    }
    buffer[0] = 0;
    buffer[1] = len;
}

/**
 * @brief   Move bytes in buffer when receiving message
 * 
 * @param   buffer      buffer with data
 */
void iCannotMoveItAnymore(char *buffer){
    int tmp = buffer[1];
    int len = buffer[2];
    if(buffer[0] == 1){
        for(int i = 0; i < len; i++){
            buffer[i] = buffer[i+3];
        }
        buffer[len+1] = 0;
        if(tmp == 0){
            printf("OK:%s\n", buffer);
        }else{
            printf("ERR:%s\n", buffer);
        }
    } 
}

/**
 * @brief   Main UDP function
 * 
 * @param   port_number         Number of wanted port
 * @param   server_hostname     Name of server
 * @return  int 
 */
int udp (int port_number, char *server_hostname) {
	int client_socket, bytestx, bytesrx;
    socklen_t serverlen;
    struct hostent *server;
    struct sockaddr_in server_address;
    char buf[BUFSIZE];
    
    //Using DNS to get the adress
    if((server = gethostbyname(server_hostname)) == NULL){
        fprintf(stderr,"ERROR: no such host as %s\n", server_hostname);
        exit(EXIT_FAILURE);
    }
    
    //Find an IP adress of the server
    bzero((char *) &server_address, sizeof(server_address));
    server_address.sin_family = AF_INET;
    bcopy((char *)server->h_addr, (char *)&server_address.sin_addr.s_addr, server->h_length);
    server_address.sin_port = htons(port_number);
    
    //Print info about server socket
    //printf("INFO: Server socket: %s : %d \n", inet_ntoa(server_address.sin_addr), ntohs(server_address.sin_port));
    
    //Make a socket
	if ((client_socket = socket(AF_INET, SOCK_DGRAM, 0)) <= 0)
	{
		perror("ERROR: socket");
		exit(EXIT_FAILURE);
	}


    //Load an user message
    while(true){
        bzero(buf, BUFSIZE);

        if(fgets(buf, BUFSIZE, stdin) == NULL){
            // If EOF is reached, break out of the loop
            if(feof(stdin)){
                break;
            }
            else{
                perror("ERROR: fgets");
                return 1;
            }
        }
    int len = strlen(buf);
        //Send a message to server
        serverlen = sizeof(server_address);
        iLikeToMoveItMoveIt(buf);
        bytestx = sendto(client_socket, buf, len + 2, 0, (struct sockaddr *) &server_address, serverlen);
        if (bytestx < 0) 
            perror("ERROR: sendto");
        
        //Receiving message from server
        bytesrx = recvfrom(client_socket, buf, BUFSIZE, 0, (struct sockaddr *) &server_address, &serverlen);
        if (bytesrx < 0) 
            perror("ERROR: recvfrom");
        iCannotMoveItAnymore(buf);

    }
    close(client_socket);
    return 0;
}
//End Of File udp.cpp

ipkcpc.cpp
/**
 * @file        ipkcpc.cpp
 * @author      Veronika Jirmusova, xjirmu00
 * @brief       Main file that activate the whole program.
 * @version     0.1
 * @date        2023-03-10
 * 
 * @copyright   Copyright (c) 2023
 * 
 */

#include "tcp.h"
#include "udp.h"
using namespace std;

/**
 * @brief   Main function.
 * 
 * @param   argc 
 * @param   argv 
 * @return  int 
 */
int main(int argc, char * argv[]){
//Arg check
    bool help_requested = false;

    // Loop through all command-line arguments
    for (int i = 1; i < argc; i++) {
        // Check if the argument is "--help"
        if (strcmp(argv[i], "--help") == 0) {
            help_requested = true;
            break;
        }
    }

    //If the "--help" flag was passed, print a help message
    if (help_requested) {
        cout << "Optional arguments:" << endl;
        cout << "  --help          Display this help message and exit." << endl;
        cout << "Mandatory arguments:" << endl;
        cout << "  -h  <host>      Where host is the IPv4 adress of the server." << endl;
        cout << "  -p  <port>      Where port is the server port." << endl;
        cout << "  -m  <mode>      Where mode is either tcp or udp." << endl;
        return 0;
    }

    //If the number of argiments is wrong and --help wasn't used, break the process
    if(argc != 7 && !help_requested){
        fprintf(stderr, "Wrong number of params! Use --help for help!\n");
        return -1;
    }

    int tmp, port;
    char *host = NULL;
    char *mode = NULL;
    //Check if all parameters are correct
    while ((tmp = getopt(argc, argv, "h:p:m:")) != -1){
        switch (tmp)
        {
        case 'h':
            host = optarg;
            break;
        case 'p':
            char *ptr;
            port = strtol(optarg, &ptr, 10);
            if(ptr[0] != 0){
                fprintf(stderr, "Wrong port argument! Use --help for help!\n");
                return -1;
            }
            if(port > 65535 || port < 0){
                fprintf(stderr, "Wrong port argument! Use --help for help!\n");
                return -1;
            }
            break;
        case 'm':
            mode = optarg;
            if(strcmp(mode, "tcp") && strcmp(mode, "udp")){
                fprintf(stderr, "Wrong mode argument! Use --help for help!\n");
                return -1;
            }
            break;
        
        default:
            fprintf(stderr, "Unknown params! Use --help for help!\n");
            return -1;
            break;
        }
    }
    
    //Start the process
    if(!strcmp(mode, "tcp")){
        tcp(port, host);
    }else{
        udp(port,host);
    }


return 0;
}
//End Of File ipkcpc.cpp